cupy/cuda/pinned_memory.pyx

```diff
@@ -271,42 +276,49 @@ cdef class PinnedMemoryPool:
         self._free = collections.defaultdict(list)
         self._alloc = allocator
         self._weakref = weakref.ref(self)
+        self._lock = rlock.create_fastrlock()
         self._allocation_unit_size = 512
 
     cpdef PinnedMemoryPointer malloc(self, Py_ssize_t size):
         cdef list free
-        cdef PinnedMemory mem
 
         if size == 0:
             return PinnedMemoryPointer(PinnedMemory(0), 0)
 
         # Round up the memory size to fit memory alignment of cudaHostAlloc
         unit = self._allocation_unit_size
-        size = (((size + unit - 1) // unit) * unit)
-        free = self._free[size]
-        if free:
-            mem = free.pop()
-        else:
-            try:
-                mem = self._alloc(size).mem
-            except runtime.CUDARuntimeError as e:
-                if e.status != runtime.errorMemoryAllocation:
-                    raise
-                self.free_all_blocks()
-                mem = self._alloc(size).mem
-
-        self._in_use[mem.ptr] = mem
+        size = ((size + unit - 1) // unit) * unit
+        rlock.lock_fastrlock(self._lock, -1, True)
+        try:
+            free = self._free[size]
+            if free:
+                mem = free.pop()
+            else:
+                try:
+                    mem = self._alloc(size).mem
+                except runtime.CUDARuntimeError as e:
+                    if e.status != runtime.errorMemoryAllocation:
+                        raise
+                    self.free_all_blocks()
+                    mem = self._alloc(size).mem
+
+            self._in_use[mem.ptr] = mem
+        finally:
+            rlock.unlock_fastrlock(self._lock)
         pmem = PooledPinnedMemory(mem, self._weakref)
         return PinnedMemoryPointer(pmem, 0)
 
     cpdef free(self, size_t ptr, Py_ssize_t size):
         cdef list free
-        cdef PinnedMemory mem
-        mem = self._in_use.pop(ptr, None)
-        if mem is None:
-            raise RuntimeError('Cannot free out-of-pool memory')
-        free = self._free[size]
-        free.append(mem)
+        rlock.lock_fastrlock(self._lock, -1, True)
+        try:
+            mem = self._in_use.pop(ptr, None)
+            if mem is None:
+                raise RuntimeError('Cannot free out-of-pool memory')
+            free = self._free[size]
+            free.append(mem)
+        finally:
+            rlock.unlock_fastrlock(self._lock)
 
     cpdef free_all_blocks(self):
         """Release free all blocks."""
```

Comment:
This is a public method, so do we need locking inside this, too?

Reply:
I fixed.